### ISSUE TITLE
`pkg build` clean?

### ISSUE THREAD

**Issue by magnusviri** [pkgxdev/brewkit#82]
Can `pkg build` clean at the start? Or add a flag to do so?

**Comment by mxcl** [comment]
I believe it should be opt-in rather than out as cleaning by mistake can waste hours of debug time.

We have `pkg clean` though it doesn’t work great. edit: well seems to work fine.

**Comment by mxcl** [comment]
next version `pkg` by itself lists available commands

**Comment by jhheider** [comment]
the main problem I have is some formulae aren't reentrant in that way: anything with a `mkdir` or a `patch`/`git apply` will fail on re-run due to targets already existing.

**Comment by mxcl** [comment]
yeah I've experienced this too.

We can have a `build.clean:` node. Specifies '*' or directories to delete when rebuilding

**Comment by mxcl** [comment]
gimme an example of something that needs this and I'll add the feature

**Comment by jhheider** [comment]
gcc, iirc. Applies a huge patch.
fish has a `mkdir build` in its instructions.

**Comment by mxcl** [comment]
patches are revoked because we always extract sources over the top. Which I agree is not exactly in line with the “we don't clean” lol.

so in the fish example we should have it specify to “clean” away the build dir?

**Comment by jhheider** [comment]
right. conversely, all build scripts can have logic like:
```sh
if test -d build; then
  # rm -rf build/*  # if you want this clean
else
  mkdir build
fi
```

**Comment by jhheider** [comment]
Or, `test -d build || mkdir build` instead of `mkdir build`. or `mkdir build || true`. Basically, for things that call `mkdir`, it should not fail if the dir exists.

scripts that cp/mv can also have this issue (target exists)

**Comment by mxcl** [comment]
So in this situation the build directory *should be preserved* to allow debuggable rebuilds. `mkdir -p build` would be an easy way to stop this from erroring.

I kinda feel that another possible solution here is `Makefile`s.

Either way I don't want to default to the super destructive “clean by default” that can cause hours of lost time. Maybe there's a better solution lurking. I'll wait for inspiration.

**Comment by mxcl** [comment]
Closing as `pkg clean && pkg build` is not too onerous.

**Comment by magnusviri** [comment]
I tried to run `pkg clean` but I get this error.

```1
/opt/tea/tea.xyz/brewkit/v0.7.1/bin/pkg: line 17: /opt/tea/tea.xyz/brewkit/v0.7.1/bin/pkg-clean: No such file or directory
```

I wasn't sure if I had the latest version so I `rm -r /opt/tea/tea.xyz/brewkit` and when I ran `pkg` it downloaded 0.7.1. I made sure I had the latest tea (0.24.10) and ran `tea -S` also.

It would be nice if I could specify a flag to clean since I usually have multiple projects and I don't want to build them all or clean them all at once. `pkg clean project-name && pkg build project-name` can get a little long when it could be `pkg -C build project-name` or something like that.

Thanks!
